pkg/controller/statefulset/stateful_set_control.go

```diff
@@ -0,0 +1,145 @@
+// implements the documented semantics for StatefulSets. podControl is the PodControlInterface used to create, update,
+// and delete Pods and to create PersistentVolumeClaims. You should use an instance returned from
+// NewRealStatefulPodControl() for any scenario other than testing.
+func NewDefaultStatefulSetControl(podControl StatefulPodControlInterface) StatefulSetControlInterface {
+	return &defaultStatefulSetControl{podControl}
+}
+
+type defaultStatefulSetControl struct {
+	podControl StatefulPodControlInterface
+}
+
+func (ssc *defaultStatefulSetControl) UpdateStatefulSet(set *apps.StatefulSet, pods []*v1.Pod) error {
+	replicaCount := int(*set.Spec.Replicas)
+	// slice that will contain all Pods such that 0 <= getOrdinal(pod) < set.Spec.Replicas
+	replicas := make([]*v1.Pod, replicaCount)
+	// slice that will contain all Pods such that set.Spec.Replicas <= getOrdinal(pod)
+	condemned := make([]*v1.Pod, 0, len(pods))
+	ready := 0
+	unhealthy := 0
+
+	// First we partition pods into two lists valid replicas and condemned Pods
+	for i := range pods {
+		//count the number of running and ready replicas
+		if isRunningAndReady(pods[i]) {
+			ready++
+		}
+		if ord := getOrdinal(pods[i]); 0 <= ord && ord < replicaCount {
+			// if the ordinal of the pod is within the range of the current number of replicas,
+			// insert it at the indirection of its ordinal
+			replicas[ord] = pods[i]
+
+		} else if ord >= replicaCount {
+			// if the ordinal is greater than the number of replicas add it to the condemned list
+			condemned = append(condemned, pods[i])
+		}
+	}
+
+	// for any empty indices in the sequence [0,set.Spec.Replicas) create a new Pod
+	for ord := 0; ord < replicaCount; ord++ {
+		if replicas[ord] == nil {
+			replicas[ord] = newStatefulSetPod(set, ord)
+		}
+	}
+
+	// count the number of unhealthy pods
+	for i := range replicas {
+		if !isHealthy(replicas[i]) {
+			unhealthy++
+		}
+	}
+	for i := range condemned {
+		if !isHealthy(condemned[i]) {
+			unhealthy++
+		}
+	}
+
+	// sort the condemned Pods by their ordinals
+	sort.Sort(ascendingOrdinal(condemned))
+
+	// if the current number of replicas has changed update the statefulSets replicas
+	if err := ssc.podControl.UpdateStatefulSetReplicas(set, int32(ready)); err != nil {
+		return err
+	}
+
+	// Examine each replica with respect to its ordinal
+	for i := range replicas {
+		// delete and recreate failed pods
+		if isFailed(replicas[i]) {
+			glog.V(2).Infof("StatefulSet %s is recreating failed Pod %s", set.Name, replicas[i].Name)
+			if err := ssc.podControl.DeleteStatefulPod(set, replicas[i]); err != nil {
+				return err
+			}
+			replicas[i] = newStatefulSetPod(set, i)
+		}
+		// If we find a Pod that has not been created we create the Pod immediately and return
+		if !isCreated(replicas[i]) {
+			return ssc.podControl.CreateStatefulPod(set, replicas[i])
+		}
+		// if we have a Pod that has been created but is not running and ready we can not make progress
```

Comment:
It's valuable to extend this comment here to include more information as to why we are doing this - with respect to the guarantee statefulsets provide to their users (ordering of pods that came up matters)